packages/jest-haste-map/src/index.js

```diff
@@ -299,30 +301,66 @@ class HasteMap extends EventEmitter {
     mocks: Object,
     filePath: Path,
     workerOptions: ?{forceInBand: boolean},
-  ): ?Promise<void> {
-    const setModule = (id: string, module: ModuleMetaData) => {
+  ): ?Promise<?{id: string, candidates: Array<ModuleMetaData>}> {
+
+    const updateModuleFromCandidates = (
+      id: string,
+      platform: string,
+      candidates: Array<ModuleMetaData>,
+    ) => {
       if (!map[id]) {
         map[id] = Object.create(null);
       }
-      const moduleMap = map[id];
-      const platform = getPlatformExtension(module[H.PATH]) ||
-        H.GENERIC_PLATFORM;
-      const existingModule = moduleMap[platform];
-      if (existingModule && existingModule[H.PATH] !== module[H.PATH]) {
-        const message = `jest-haste-map: @providesModule naming collision:\n` +
-          `  Duplicate module name: ${id}\n` +
-          `  Paths: ${module[H.PATH]} collides with ` +
-          `${existingModule[H.PATH]}\n\nThis ` +
-          `${this._options.throwOnModuleCollision ? 'error' : 'warning'} ` +
-          `is caused by a @providesModule declaration ` +
-          `with the same name across two different files.`;
-        if (this._options.throwOnModuleCollision) {
-          throw new Error(message);
+      const modulesByPlatform = map[id];
+      if (candidates.length === 1) {
+        modulesByPlatform[platform] = candidates[0];
+      } else {
+        delete modulesByPlatform[platform];
```

Comment:
I think it may make sense to add a comment here:

```
// If there are more than one candidate for the same module, we remove them from the map.
```